flink-contrib/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBListState.java

```diff
@@ -167,24 +167,48 @@ public void update(List<V> values) throws Exception {
 			try {
 				writeCurrentKeyWithGroupAndNamespace();
 				byte[] key = keySerializationStream.toByteArray();
-				DataOutputViewStreamWrapper out = new DataOutputViewStreamWrapper(keySerializationStream);
 
-				List<byte[]> bytes = new ArrayList<>(values.size());
-				for (V value : values) {
-					keySerializationStream.reset();
-					valueSerializer.serialize(value, out);
-					bytes.add(keySerializationStream.toByteArray());
+				byte[] premerge = getPreMergedValue(values);
+				if (premerge != null) {
+					backend.db.put(columnFamily, writeOptions, key, premerge);
+				} else {
+					throw new IOException("Failed pre-merge values in update()");
 				}
+			} catch (IOException | RocksDBException e) {
+				throw new RuntimeException("Error while updating data to RocksDB", e);
+			}
+		}
+	}
+
+	@Override
+	public void addAll(List<V> values) throws Exception {
+		if (values != null && !values.isEmpty()) {
+			try {
+				writeCurrentKeyWithGroupAndNamespace();
+				byte[] key = keySerializationStream.toByteArray();
 
-				byte[] premerge = MergeUtils.merge(bytes);
+				byte[] premerge = getPreMergedValue(values);
 				if (premerge != null) {
-					backend.db.put(columnFamily, writeOptions, key, premerge);
+					backend.db.merge(columnFamily, writeOptions, key, premerge);
 				} else {
-					throw new IOException("Failed pre-merge values");
+					throw new IOException("Failed pre-merge values in addAll()");
 				}
 			} catch (IOException | RocksDBException e) {
 				throw new RuntimeException("Error while updating data to RocksDB", e);
 			}
 		}
 	}
+
+	private byte[] getPreMergedValue(List<V> values) throws IOException {
+		DataOutputViewStreamWrapper out = new DataOutputViewStreamWrapper(keySerializationStream);
+
+		List<byte[]> bytes = new ArrayList<>(values.size());
```

Comment:
Why not just serialize all objects in the stream and write the RocksDB separator byte between all object bytes? This could improve performance (less temporary copies and objects) and might be equally or more readable?

From a performance point of view, even `#toByteArray()` results in an unnecessary copy - we could use the internal array, offset + len for our insert to RocksDB.

What do you think?

Reply:
@StefanRRichter  I actually tried it before, but it didn't work out very well. I can give it another try.

I don't think this PR should be addressing this issue, because that code is already there and this PR only move it to its own method. Besides, it will be great if we can get the new API in before I announce it in our Flink meetup at Seattle's Wednesday evening :) (Thanks in advance if that may take extra work from your end!)

I opened FLINK-8441 and I'll be working on it shortly after.

Reply:
Interesting, do you have an idea why it did not work? I think it should be possible. In general, I am not a big fan of changing this code twice when we already assume that we do an overhaul of that part, but we can do it for this time if it makes your life easier.

Sorry that we could not get it done for the meetup, but I was blocked with another important matter :-(

Reply:
That'll be great, let's get this in. I will dive into FLINK-8441 in a couple days. Thanks!